File: src/scaffolds/backends/go/services/file_service.go
// Package services 文件服务
package services

import (
	"fmt"
	"go-mysql-backend/config"
	"go-mysql-backend/utils"
	"mime/multipart"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

var allowedExtensions = map[string]bool{
	".jpg": true, ".jpeg": true, ".png": true, ".gif": true, ".bmp": true, ".webp": true,
	".pdf": true, ".doc": true, ".docx": true, ".xls": true, ".xlsx": true, ".ppt": true, ".pptx": true,
	".txt": true, ".zip": true, ".rar": true, ".7z": true,
}

// FileUpload 上传单个文件
func FileUpload(c *gin.Context, file *multipart.FileHeader) (map[string]interface{}, error) {
	if err := validateFile(file); err != nil {
		return nil, err
	}
	url, err := saveFile(c, file)
	if err != nil {
		return nil, err
	}
	return map[string]interface{}{"url": url, "fileName": file.Filename}, nil
}

// FileUploadBatch 批量上传文件
func FileUploadBatch(c *gin.Context, files []*multipart.FileHeader) map[string]interface{} {
	success := []map[string]string{}
	fail := []string{}

	for _, file := range files {
		if err := validateFile(file); err != nil {
			fail = append(fail, fmt.Sprintf("%s: %s", file.Filename, err.Error()))
			continue
		}
		url, err := saveFile(c, file)
		if err != nil {
			fail = append(fail, fmt.Sprintf("%s: %s", file.Filename, err.Error()))
			continue
		}
		success = append(success, map[string]string{"fileName": file.Filename, "url": url})
	}

	return map[string]interface{}{
		"success": success, "fail": fail,
		"total": len(files), "successCount": len(success), "failCount": len(fail),
	}
}

// FileDelete 删除文件
func FileDelete(fileName string) error {
	if fileName == "" {
		return &BizError{Code: utils.CodeBadRequest, Message: "文件名不能为空"}
	}
	relativePath := strings.Replace(fileName, "/uploads/", "", 1)
	uploadConfig := config.GetUploadConfig()
	basePath, _ := filepath.Abs(uploadConfig.UploadDir)
	absolutePath, _ := filepath.Abs(filepath.Join(basePath, relativePath))
	if !strings.HasPrefix(absolutePath, basePath+string(os.PathSeparator)) && absolutePath != basePath {
		return &BizError{Code: utils.CodeBadRequest, Message: "非法的文件路径"}
	}
	if _, err := os.Stat(absolutePath); err == nil {
		os.Remove(absolutePath)
	}
	return nil
}

func validateFile(file *multipart.FileHeader) error {
	if file == nil || file.Size == 0 {
		return &BizError{Code: utils.CodeBadRequest, Message: "文件不能为空"}
	}
	uploadConfig := config.GetUploadConfig()
	if file.Size > uploadConfig.MaxFileSize {
		return &BizError{Code: utils.CodeBadRequest, Message: "文件大小不能超过10MB"}
	}
	ext := strings.ToLower(filepath.Ext(file.Filename))
	if !allowedExtensions[ext] {
		return &BizError{Code: utils.CodeBadRequest, Message: fmt.Sprintf("不支持的文件类型: %s", ext)}
	}
	return nil
}

func saveFile(c *gin.Context, file *multipart.FileHeader) (string, error) {
	ext := strings.ToLower(filepath.Ext(file.Filename))
	newName := uuid.New().String() + ext
	dateDir := time.Now().Format("2006-01-02")
	uploadConfig := config.GetUploadConfig()
	dir := filepath.Join(uploadConfig.UploadDir, dateDir)

	if err := os.MkdirAll(dir, 0755); err != nil {
		return "", &BizError{Code: utils.CodeError, Message: "创建目录失败"}
	}

	dst := filepath.Join(dir, newName)
	if err := c.SaveUploadedFile(file, dst); err != nil {
		return "", &BizError{Code: utils.CodeError, Message: "保存文件失败"}
	}

	return fmt.Sprintf("/uploads/%s/%s", dateDir, newName), nil
}

File: src/scaffolds/backends/dotnet/Services/FileService.cs
using DotnetMysqlBackend.Config;
using DotnetMysqlBackend.Exceptions;

namespace DotnetMysqlBackend.Services
{
    /// <summary>
    /// 文件服务
    /// </summary>
    public class FileService
    {
        private readonly UploadConfig _config;

        private static readonly HashSet<string> AllowedExtensions = new(StringComparer.OrdinalIgnoreCase)
        {
            ".jpg", ".jpeg", ".png", ".gif", ".bmp", ".webp",
            ".pdf", ".doc", ".docx", ".xls", ".xlsx", ".ppt", ".pptx",
            ".txt", ".zip", ".rar", ".7z"
        };

        public FileService(UploadConfig config)
        {
            _config = config;
        }

        /// <summary>
        /// 上传单个文件
        /// </summary>
        public async Task<object> Upload(IFormFile file)
        {
            Validate(file);
            var url = await Save(file);
            return new { url, fileName = file.FileName };
        }

        /// <summary>
        /// 批量上传文件
        /// </summary>
        public async Task<object> UploadBatch(IFormFileCollection files)
        {
            if (files == null || files.Count == 0)
                throw new BusinessException("请选择要上传的文件");

            var success = new List<object>();
            var fail = new List<string>();

            foreach (var file in files)
            {
                try
                {
                    Validate(file);
                    var url = await Save(file);
                    success.Add(new { fileName = file.FileName, url });
                }
                catch (Exception ex)
                {
                    fail.Add($"{file.FileName}: {ex.Message}");
                }
            }

            return new { success, fail, total = files.Count, successCount = success.Count, failCount = fail.Count };
        }

        /// <summary>
        /// 删除文件
        /// </summary>
        public void Delete(string fileName)
        {
            if (string.IsNullOrWhiteSpace(fileName))
                throw new BusinessException("文件名不能为空");

            var relativePath = fileName.Replace("/uploads/", "").Replace("\\", "/");
            var basePath = Path.GetFullPath(_config.UploadDir);
            var absolutePath = Path.GetFullPath(Path.Combine(basePath, relativePath));

            if (!absolutePath.StartsWith(basePath + Path.DirectorySeparatorChar) && absolutePath != basePath)
                throw new BusinessException("非法的文件路径");

            if (File.Exists(absolutePath))
                File.Delete(absolutePath);
        }

        #region 私有方法

        /// <summary>
        /// 校验文件
        /// </summary>
        private void Validate(IFormFile file)
        {
            if (file == null || file.Length == 0)
                throw new BusinessException("文件不能为空");
            if (file.Length > _config.MaxFileSize)
                throw new BusinessException("文件大小不能超过10MB");
            var ext = Path.GetExtension(file.FileName);
            if (string.IsNullOrEmpty(ext) || !AllowedExtensions.Contains(ext))
                throw new BusinessException($"不支持的文件类型: {ext}");
        }

        /// <summary>
        /// 保存文件到磁盘
        /// </summary>
        private async Task<string> Save(IFormFile file)
        {
            var ext = Path.GetExtension(file.FileName);
            var newName = $"{Guid.NewGuid():N}{ext}";
            var dateDir = DateTime.Now.ToString("yyyy-MM-dd");
            var dir = Path.GetFullPath(Path.Combine(_config.UploadDir, dateDir));

            if (!Directory.Exists(dir))
                Directory.CreateDirectory(dir);

            var path = Path.Combine(dir, newName);
            await using var stream = new FileStream(path, FileMode.Create);
            await file.CopyToAsync(stream);

            return $"/uploads/{dateDir}/{newName}";
        }

        #endregion
    }
}
